src/common/snippets/src/lowered/expression_factory.cpp

```diff
@@ -10,22 +10,29 @@ namespace ov {
 namespace snippets {
 namespace lowered {
 
-void LinearIR::ExpressionFactory::create_expression_inputs(const LinearIR& linear_ir, const ExpressionPtr& expr) {
-    OPENVINO_ASSERT(expr != nullptr, "Failed expression inputs creation: expression is null");
-    const auto& node = expr->get_node();
-
-    expr->m_input_port_connectors.resize(node->get_input_size(), nullptr);
-    for (const auto& input : node->inputs()) {
-        const auto input_source = input.get_source_output();
-        const auto in_index = input.get_index();
-        const auto& parent_expr = linear_ir.get_expr_by_node(input_source.get_node_shared_ptr());
-        const auto& port_connector = parent_expr->get_output_port_connector(input_source.get_index());
-        port_connector->add_consumer(expr->get_input_port(in_index));
-        expr->m_input_port_connectors[in_index] = port_connector;
+template<>
+std::shared_ptr<Expression> ExpressionFactory::build(const std::shared_ptr<Node>& n, const std::vector<PortConnectorPtr>& inputs) {
+    if (const auto par = ov::as_type_ptr<ov::op::v0::Parameter>(n)) {
+        return create(par, inputs, m_shape_infer_factory);
+    } else if (const auto res = ov::as_type_ptr<ov::op::v0::Result>(n)) {
+        return create(res, inputs, m_shape_infer_factory);
+    } else if (const auto loop_begin = ov::as_type_ptr<op::LoopBegin>(n)) {
+        return create(loop_begin, inputs, m_shape_infer_factory);
+    } else if (const auto loop_end = ov::as_type_ptr<op::LoopEnd>(n)) {
+        return create(loop_end, inputs, m_shape_infer_factory);
+    } else if (const auto buffer = ov::as_type_ptr<op::Buffer>(n)) {
+        return create<BufferExpression>(buffer, inputs, m_shape_infer_factory);
+#ifdef SNIPPETS_DEBUG_CAPS
+    } else if (const auto perf_counter = ov::as_type_ptr<op::PerfCountBeginBase>(n)) {
+        return create(perf_counter, inputs, m_shape_infer_factory);
+    } else if (const auto perf_counter = ov::as_type_ptr<op::PerfCountEndBase>(n)) {
+        return create(perf_counter, inputs, m_shape_infer_factory);
+#endif
     }
+    return create<>(n, inputs, m_shape_infer_factory);
```

Comment:
Only [since c++17](https://stackoverflow.com/a/50970942) we can skip `<>` . As far as I know, we should support [Centos 7](https://docs.openvino.ai/2024/about-openvino/release-notes-openvino/system-requirements.html) but I'm not sure that this OS uses c++17. I removed `<>` everywhere so let's check CI
Thanks!